src/core/inductive.ml

```diff
@@ -0,0 +1,280 @@
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), l) ->
+                let b =
+                  (* If the current symbol is equal to the inductive type,
+                     we create an inductive hypothesis. *)
+                  if s == sind then
+                    let l = List.map lift l in
+                    _Impl (prf_of_p l (Bindlib.box_var x)) b
+                  else
+                    b
+                in
+                let res = _Prod (lift a) (Bindlib.bind_var x b) in
+                if !log_enabled then
+                  log_ind "The clause of the constructor [%a] is %a"
+                    Print.pp_term (Symb(scons))
+                    Print.pp_term (Bindlib.unbox res);
+                res
+            | _ -> fatal pos "The type of %a is not supported" pp_symbol scons
+          end
+      | _ -> fatal pos "The type of %a is not supported" pp_symbol scons
+    in
+    case [] !(scons.sym_type)
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let to_create_end  : tbox list -> tbox = fun ts ->
+    let x = Bindlib.new_var mkfree "x"                 in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp sind ts) t
+  in
+  let conclusion = create_instance_of pos sind to_create_end  in
+
+  (* STEP 4: Create the inductive principle *)
+  let rec add_case l r =
+    match l with
+    | []   -> r
+    | t::q -> let t = case_of t in
+              _Impl t (add_case q r)
+  in
+  let t = add_case scons_list conclusion      in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Print.pp_term (Symb(sind)) Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [type_to_pattern t] creates a pattern which matches with the type [t]. *)
+let rec type_to_pattern : term -> p_term = fun t ->
+  match t with
+  | Vari x    -> Pos.none (P_Patt (Some (Pos.none (Bindlib.name_of x)), None))
+  | Symb s       -> Pos.none (P_Iden (Pos.none ([], s.sym_name), true))
+  | Appl(t1, t2) -> Pos.none (P_Appl (type_to_pattern t1, type_to_pattern t2))
+  | Prod _       -> fatal None "Prod - Not yet implemented"
+  | Abst _       -> fatal None "Abst - Not yet implemented"
+  | LLet _       -> fatal None "LLet - Not yet implemented"
+  | Type         -> fatal None "TYPE - Not yet implemented"(*Pos.none P_Type*)
+  | Kind         -> fatal None "[Kind] not possible"
+  | Wild         -> fatal None "[Wild] not possible"
+  | TRef _       -> fatal None "[TRef] not possible"
+  | Meta _       -> fatal None "[Meta] not possible"
+  | Patt _       -> fatal None "[Patt] not possible"
+  | TEnv _       -> fatal None "[TEnv] not possible"
+
+(** Some constructors to create a p_term without position *)
+
+let _P_Iden : qident -> bool -> p_term =
```

Comment:
do not use _ as prefix as it is already used for building tbox in terms.ml
write instead p_iden, p_patt, etc.

Reply:
style: put fun on previous line whenever possible

Reply:
Hmm it will be difficult to see if it's a function, a type, a variable or a constructor.
It's not confused with `term` because you have the uppercase `P` to highlight that it's a `p_term`.
And in a sense, it's a "smart" constructor too.

Reply:
Hmm it will be difficult to see if it's a function, a type, a variable or a constructor.
It's not confused with `term` because you have the uppercase `P` to highlight that it's a `p_term`.
And in a sense, it's a "smart" constructor too.

Reply:
No, there will be no confusion. It's usual to use the same name with lowercase letter for private constructors.